arch: riscv: fatal: always print `mcause` & `mtval`, make `cause_str()` reusable

ycsin opened the pull request:
Relocate the logging of mcause & mtval from `_Fault` to
`z_riscv_fatal_error_csf` so that they are always printed
upon exception.

Rename `z_riscv_cause_str` to `cause_str` and make it non-static,
so that it can be used in user-implemented `k_sys_fatal_error_handler`.

Also, this function should return a constant string, so add `const`
to it.

ycsin committed:
arch: riscv: fatal: always print mcause & mtval

Relocate the logging of mcause & mtval from `_Fault` to
`z_riscv_fatal_error_csf` so that they are always printed
upon exception.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
arch: riscv: fatal: make `cause_str` reusable

Rename `cause_str` to `z_riscv_mcause_str` and make it non-static,
so that it can be used in user-implemented `k_sys_fatal_error_handler`.

Also, this function should return a constant string, so add `const`
to it.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

npitre commented:
While at it, you could get rid of the switch statement and put those strings
in an array indexed with the cause value. That would reduce code size.

ycsin committed:
arch: riscv: use string array instead of switch statement for cause

Get rid of the switch statement and use an string array
for the cause instead. This saves about ~600 bytes.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
> While at it, you could get rid of the switch statement and put those strings in an array indexed with the cause value. That would reduce code size.

I added another commit for this, did a little test and seems like it saved about 100+ bytes

```
west build -b qemu_riscv64 -p always zephyr/samples/hello_world -- -DCONFIG_SYMTAB=y -DCONFIG_SHELL=y -DCONFIG_LOG=y -DCONFIG_ARCH_STACKWALK=y -DCONFIG_FRAME_POINTER=y
ls -l build/zephyr/zephyr.elf
```

```
before
-rwxr-xr-x@ 1 ycsin  staff  1492664 Sep  5 12:50 build/zephyr/zephyr.elf

after
-rwxr-xr-x@ 1 ycsin  staff  1492512 Sep  5 12:51 build/zephyr/zephyr.elf
```

npitre commented:
Nice improvement with the array. But it can be improved further.

- Define the array as static as in `static const char *mcause_str[16]`
otherwise the array will be reconstructed at run tine on the stack each
time the function is called. Bad for runtime, bad for stack usage.

- The array contains const strings, but the array itself may be const as in
`static const char * const mcause_str[16]`. This will move the array from
`.data` to `.rodata` linker section which is better for XIP among other
reasons.

- Finally (and you need to be a disassembly freak to notice that) it is more
efficient to add a 17th entry with "unknown" and cap the index to <= 16
than returning "unknown" directly in case of overflow.

With the above you should save yet more bytes.

ycsin commented:
> With the above you should save yet more bytes.

I followed all that but instead of adding 17th entry, I set the cause to `14` so that "unknown" is returned without having an additional entry, and now the difference is -740 bytes:

```
$ ls -l build/zephyr/zephyr.elf
-rwxr-xr-x 1 ycsin ycsin 1491904 Sep  6 00:41 build/zephyr/zephyr.elf
```

Thanks for the suggestion and the detailed info!

npitre commented:
Well... hardcoding 14 when overflow is maybe a tad too far though.

You saved many bytes already, so I'd suggest making the 17th entry 
"invalid cause value" and then you could simply do:

```
    return mcause_str[MIN(cause, ARRAY_SIZE(mcause_str) - 1)];
```
